Can we show the source command for all error messages?

Consider the below program:

```javascript
const regl = require('../regl')()

regl( {
  frag: [
    'precision mediump float;',
    'uniform vec4 color;',
    'void main() {',
    '  gl_FragColor = vec4(1, 0, 0, 1);',
    '}'
  ].join('\n'),

  vert: [
    'precision mediump float;',
    'attribute vec2 position;',
    'void main() {',
    '  gl_Position = vec4(position, 0, 1);',
    '}'
  ].join('\n'),

  attributes: {
    position: regl.buffer([
      [0, 0],
      [1, 0],
      [1, 1],
      [0, 1],
      [1, 0],
      [1, 1]
    ])
  },

  count: 6,
  blend:{
    equation: {
      rgb: 'reverse subtract',
      alpha: 'add'
    },
  },
  stencil: {
    enable: true,
    mask: 0xff,
  }
})()
```

If we comment out the fragment shader, we get a very useful error message

```
check.js:21Error: (regl) missing fragment shader in command http://130.241.189.159:9966/bug2.js:4:1 
called from http://130.241.189.159:9966/bug2.js:43:3(…)
```

this useful error message shows you the exact command that causes the issue. However, if you for instance set

```javascript
      alpha: 'adddddd'
```

in the above program, you don't get a very good error message

```
check.js:21Error: (regl) unknown parameter (addddd): blend.equation.alpha. 
possible values: add,subtract,reverse subtract in command unknown(…)
```

this error message is bad, because it doesn't show you the source command. You also get that kind of error message if you set

```
    mask: 'bla',

```

for the stencil buffer.

So my question: is it possible to implement so that we always get to know the source command for all error messages? For some error messages we get to know the source command, but it is far from all. But knowing the source command could be very useful when debugging big code bases.

 This would make regl more pleasant to use, I think.

--- a/lib/core.js
+++ b/lib/core.js
@@ -570,7 +570,7 @@ module.exports = function reglCore (
     return profileEnable
   }
 
-  function parseFramebuffer (options) {
+  function parseFramebuffer (options, env) {
     var staticOptions = options.static
     var dynamicOptions = options.dynamic
 
@@ -654,29 +654,29 @@ module.exports = function reglCore (
     }
   }
 
-  function parseViewportScissor (options, framebuffer) {
+  function parseViewportScissor (options, framebuffer, env) {
     var staticOptions = options.static
     var dynamicOptions = options.dynamic
 
     function parseBox (param) {
       if (param in staticOptions) {
         var box = staticOptions[param]
-        check.commandType(box, 'object', 'invalid ' + param)
+        check.commandType(box, 'object', 'invalid ' + param, env.commandStr)
 
         var isStatic = true
         var x = box.x | 0
         var y = box.y | 0
-        check.command(x >= 0 && y >= 0, 'invalid ' + param)
+        check.command(x >= 0 && y >= 0, 'invalid ' + param, env.commandStr)
         var w, h
         if ('width' in box) {
           w = box.width | 0
-          check.command(w > 0, 'invalid ' + param)
+          check.command(w > 0, 'invalid ' + param, env.commandStr)
         } else {
           isStatic = false
         }
         if ('height' in box) {
           h = box.height | 0
-          check.command(h > 0, 'invalid ' + param)
+          check.command(h > 0, 'invalid ' + param, env.commandStr)
         } else {
           isStatic = false
         }
@@ -691,13 +691,13 @@ module.exports = function reglCore (
             if (!('width' in box)) {
               BOX_W = scope.def(CONTEXT, '.', S_FRAMEBUFFER_WIDTH, '-', x)
             } else {
-              check.command(w > 0, 'invalid ' + param)
+              check.command(w > 0, 'invalid ' + param, env.commandStr)
             }
             var BOX_H = h
             if (!('height' in box)) {
               BOX_H = scope.def(CONTEXT, '.', S_FRAMEBUFFER_HEIGHT, '-', y)
             } else {
-              check.command(h > 0, 'invalid ' + param)
+              check.command(h > 0, 'invalid ' + param, env.commandStr)
             }
             return [x, y, BOX_W, BOX_H]
           })
@@ -863,7 +863,7 @@ module.exports = function reglCore (
     }
   }
 
-  function parseDraw (options) {
+  function parseDraw (options, env) {
     var staticOptions = options.static
     var dynamicOptions = options.dynamic
 
@@ -874,7 +874,7 @@ module.exports = function reglCore (
           elements = elementState.getElements(elementState.create(elements, true))
         } else if (elements) {
           elements = elementState.getElements(elements)
-          check.command(elements, 'invalid elements')
+          check.command(elements, 'invalid elements', env.commandStr)
         }
         var result = createStaticDecl(function (env, scope) {
           if (elements) {
@@ -928,7 +928,7 @@ module.exports = function reglCore (
     function parsePrimitive () {
       if (S_PRIMITIVE in staticOptions) {
         var primitive = staticOptions[S_PRIMITIVE]
-        check.commandParameter(primitive, primTypes, 'invalid primitve')
+        check.commandParameter(primitive, primTypes, 'invalid primitve', env.commandStr)
         return createStaticDecl(function (env, scope) {
           return primTypes[primitive]
         })
@@ -972,7 +972,7 @@ module.exports = function reglCore (
     function parseParam (param, isOffset) {
       if (param in staticOptions) {
         var value = staticOptions[param] | 0
-        check.command(!isOffset || value >= 0, 'invalid ' + param)
+        check.command(!isOffset || value >= 0, 'invalid ' + param, env.commandStr)
         return createStaticDecl(function (env, scope) {
           if (isOffset) {
             env.OFFSET = value
@@ -1008,7 +1008,7 @@ module.exports = function reglCore (
       if (S_COUNT in staticOptions) {
         var count = staticOptions[S_COUNT] | 0
         check.command(
-          typeof count === 'number' && count >= 0, 'invalid vertex count')
+          typeof count === 'number' && count >= 0, 'invalid vertex count', env.commandStr)
         return createStaticDecl(function () {
           return count
         })
@@ -1090,7 +1090,7 @@ module.exports = function reglCore (
     }
   }
 
-  function parseGLState (options) {
+  function parseGLState (options, env) {
     var staticOptions = options.static
     var dynamicOptions = options.dynamic
 
@@ -1126,14 +1126,14 @@ module.exports = function reglCore (
         case S_DEPTH_MASK:
           return parseParam(
             function (value) {
-              check.commandType(value, 'boolean', prop)
+              check.commandType(value, 'boolean', prop, env.commandStr)
               return value
             },
             function (env, scope, value) {
               check.optional(function () {
                 env.assert(scope,
                   'typeof ' + value + '==="boolean"',
-                  'invalid flag ' + prop)
+                  'invalid flag ' + prop, env.commandStr)
               })
               return value
             })
@@ -1141,7 +1141,7 @@ module.exports = function reglCore (
         case S_DEPTH_FUNC:
           return parseParam(
             function (value) {
-              check.commandParameter(value, compareFuncs, 'invalid ' + prop)
+              check.commandParameter(value, compareFuncs, 'invalid ' + prop, env.commandStr)
               return compareFuncs[value]
             },
             function (env, scope, value) {
@@ -1163,7 +1163,8 @@ module.exports = function reglCore (
                 typeof value[0] === 'number' &&
                 typeof value[1] === 'number' &&
                 value[0] <= value[1],
-                'depth range is 2d array')
+                'depth range is 2d array',
+                env.commandStr)
               return value
             },
             function (env, scope, value) {
@@ -1185,19 +1186,19 @@ module.exports = function reglCore (
         case S_BLEND_FUNC:
           return parseParam(
             function (value) {
-              check.commandType(value, 'object', 'blend.func')
+              check.commandType(value, 'object', 'blend.func', env.commandStr)
               var srcRGB = ('srcRGB' in value ? value.srcRGB : value.src)
               var srcAlpha = ('srcAlpha' in value ? value.srcAlpha : value.src)
               var dstRGB = ('dstRGB' in value ? value.dstRGB : value.dst)
               var dstAlpha = ('dstAlpha' in value ? value.dstAlpha : value.dst)
-              check.commandParameter(srcRGB, blendFuncs, param + '.srcRGB')
-              check.commandParameter(srcAlpha, blendFuncs, param + '.srcAlpha')
-              check.commandParameter(dstRGB, blendFuncs, param + '.dstRGB')
-              check.commandParameter(dstAlpha, blendFuncs, param + '.dstAlpha')
+              check.commandParameter(srcRGB, blendFuncs, param + '.srcRGB', env.commandStr)
+              check.commandParameter(srcAlpha, blendFuncs, param + '.srcAlpha', env.commandStr)
+              check.commandParameter(dstRGB, blendFuncs, param + '.dstRGB', env.commandStr)
+              check.commandParameter(dstAlpha, blendFuncs, param + '.dstAlpha', env.commandStr)
 
               check.command(
                 (invalidBlendCombinations.indexOf(srcRGB + ', ' + dstRGB) === -1),
-                'unallowed blending combination (srcRGB, dstRGB) = (' + srcRGB + ', ' + dstRGB + ')')
+                'unallowed blending combination (srcRGB, dstRGB) = (' + srcRGB + ', ' + dstRGB + ')', env.commandStr)
 
               return [
                 blendFuncs[srcRGB],
@@ -1255,22 +1256,22 @@ module.exports = function reglCore (
           return parseParam(
             function (value) {
               if (typeof value === 'string') {
-                check.commandParameter(value, blendEquations, 'invalid ' + prop)
+                check.commandParameter(value, blendEquations, 'invalid ' + prop, env.commandStr)
                 return [
                   blendEquations[value],
                   blendEquations[value]
                 ]
               } else if (typeof value === 'object') {
                 check.commandParameter(
-                  value.rgb, blendEquations, prop + '.rgb')
+                  value.rgb, blendEquations, prop + '.rgb', env.commandStr)
                 check.commandParameter(
-                  value.alpha, blendEquations, prop + '.alpha')
+                  value.alpha, blendEquations, prop + '.alpha', env.commandStr)
                 return [
                   blendEquations[value.rgb],
                   blendEquations[value.alpha]
                 ]
               } else {
-                check.commandRaise('invalid blend.equation')
+                check.commandRaise('invalid blend.equation', env.commandStr)
               }
             },
             function (env, scope, value) {
@@ -1313,7 +1314,7 @@ module.exports = function reglCore (
               check.command(
                 isArrayLike(value) &&
                 value.length === 4,
-                'blend.color must be a 4d array')
+                'blend.color must be a 4d array', env.commandStr)
               return loop(4, function (i) {
                 return +value[i]
               })
@@ -1333,7 +1334,7 @@ module.exports = function reglCore (
         case S_STENCIL_MASK:
           return parseParam(
             function (value) {
-              check.commandType(value, 'number', param)
+              check.commandType(value, 'number', param, env.commandStr)
               return value | 0
             },
             function (env, scope, value) {
@@ -1348,13 +1349,13 @@ module.exports = function reglCore (
         case S_STENCIL_FUNC:
           return parseParam(
             function (value) {
-              check.commandType(value, 'object', param)
+              check.commandType(value, 'object', param, env.commandStr)
               var cmp = value.cmp || 'keep'
               var ref = value.ref || 0
               var mask = 'mask' in value ? value.mask : -1
-              check.commandParameter(cmp, compareFuncs, prop + '.cmp')
-              check.commandType(ref, 'number', prop + '.ref')
-              check.commandType(mask, 'number', prop + '.mask')
+              check.commandParameter(cmp, compareFuncs, prop + '.cmp', env.commandStr)
+              check.commandType(ref, 'number', prop + '.ref', env.commandStr)
+              check.commandType(mask, 'number', prop + '.mask', env.commandStr)
               return [
                 compareFuncs[cmp],
                 ref,
@@ -1388,13 +1389,13 @@ module.exports = function reglCore (
         case S_STENCIL_OPBACK:
           return parseParam(
             function (value) {
-              check.commandType(value, 'object', param)
+              check.commandType(value, 'object', param, env.commandStr)
               var fail = value.fail || 'keep'
               var zfail = value.zfail || 'keep'
               var pass = value.pass || 'keep'
-              check.commandParameter(fail, stencilOps, prop + '.fail')
-              check.commandParameter(zfail, stencilOps, prop + '.zfail')
-              check.commandParameter(pass, stencilOps, prop + '.pass')
+              check.commandParameter(fail, stencilOps, prop + '.fail', env.commandStr)
+              check.commandParameter(zfail, stencilOps, prop + '.zfail', env.commandStr)
+              check.commandParameter(pass, stencilOps, prop + '.pass', env.commandStr)
               return [
                 prop === S_STENCIL_OPBACK ? GL_BACK : GL_FRONT,
                 stencilOps[fail],
@@ -1436,11 +1437,11 @@ module.exports = function reglCore (
         case S_POLYGON_OFFSET_OFFSET:
           return parseParam(
             function (value) {
-              check.commandType(value, 'object', param)
+              check.commandType(value, 'object', param, env.commandStr)
               var factor = value.factor | 0
               var units = value.units | 0
-              check.commandType(factor, 'number', param + '.factor')
-              check.commandType(units, 'number', param + '.units')
+              check.commandType(factor, 'number', param + '.factor', env.commandStr)
+              check.commandType(units, 'number', param + '.units', env.commandStr)
               return [factor, units]
             },
             function (env, scope, value) {
@@ -1465,7 +1466,7 @@ module.exports = function reglCore (
               } else if (value === 'back') {
                 face = GL_BACK
               }
-              check.command(!!face, param)
+              check.command(!!face, param, env.commandStr)
               return face
             },
             function (env, scope, value) {
@@ -1486,7 +1487,7 @@ module.exports = function reglCore (
                 value >= limits.lineWidthDims[0] &&
                 value <= limits.lineWidthDims[1],
                 'invalid line width, must positive number between ' +
-                limits.lineWidthDims[0] + ' and ' + limits.lineWidthDims[1])
+                limits.lineWidthDims[0] + ' and ' + limits.lineWidthDims[1], env.commandStr)
               return value
             },
             function (env, scope, value) {
@@ -1504,7 +1505,7 @@ module.exports = function reglCore (
         case S_FRONT_FACE:
           return parseParam(
             function (value) {
-              check.commandParameter(value, orientationType, param)
+              check.commandParameter(value, orientationType, param, env.commandStr)
               return orientationType[value]
             },
             function (env, scope, value) {
@@ -1522,7 +1523,7 @@ module.exports = function reglCore (
             function (value) {
               check.command(
                 isArrayLike(value) && value.length === 4,
-                'color.mask must be length 4 array')
+                'color.mask must be length 4 array', env.commandStr)
               return value.map(function (v) { return !!v })
             },
             function (env, scope, value) {
@@ -1540,13 +1541,13 @@ module.exports = function reglCore (
         case S_SAMPLE_COVERAGE:
           return parseParam(
             function (value) {
-              check.command(typeof value === 'object' && value, param)
+              check.command(typeof value === 'object' && value, param, env.commandStr)
               var sampleValue = 'value' in value ? value.value : 1
               var sampleInvert = !!value.invert
               check.command(
                 typeof sampleValue === 'number' &&
                 sampleValue >= 0 && sampleValue <= 1,
-                'sample.coverage.value must be a number between 0 and 1')
+                'sample.coverage.value must be a number between 0 and 1', env.commandStr)
               return [sampleValue, sampleInvert]
             },
             function (env, scope, value) {
@@ -1566,65 +1567,7 @@ module.exports = function reglCore (
     return STATE
   }
 
-  function parseOptions (options) {
-    var staticOptions = options.static
-    var dynamicOptions = options.dynamic
-
-    check.optional(function () {
-      var command = check.guessCommand()
-
-      var KEY_NAMES = [
-        S_FRAMEBUFFER,
-        S_VERT,
-        S_FRAG,
-        S_ELEMENTS,
-        S_PRIMITIVE,
-        S_OFFSET,
-        S_COUNT,
-        S_INSTANCES,
-        S_PROFILE
-      ].concat(GL_STATE_NAMES)
-
-      function checkKeys (dict) {
-        Object.keys(dict).forEach(function (key) {
-          check.command(
-            KEY_NAMES.indexOf(key) >= 0,
-            'unknown parameter "' + key + '"',
-            command)
-        })
-      }
-
-      checkKeys(staticOptions)
-      checkKeys(dynamicOptions)
-    })
-
-    var framebuffer = parseFramebuffer(options)
-    var viewportAndScissor = parseViewportScissor(options, framebuffer)
-    var draw = parseDraw(options)
-    var state = parseGLState(options)
-    var shader = parseProgram(options)
-
-    function copyBox (name) {
-      var defn = viewportAndScissor[name]
-      if (defn) {
-        state[name] = defn
-      }
-    }
-    copyBox(S_VIEWPORT)
-    copyBox(propName(S_SCISSOR_BOX))
-
-    var dirty = Object.keys(state).length > 0
-
-    return {
-      framebuffer: framebuffer,
-      draw: draw,
-      shader: shader,
-      state: state,
-      dirty: dirty
-    }
-  }
-
-  function parseUniforms (uniforms) {
+  function parseUniforms (uniforms, env) {
     var staticUniforms = uniforms.static
     var dynamicUniforms = uniforms.dynamic
 
@@ -1647,13 +1590,13 @@ module.exports = function reglCore (
           })
         } else if (reglType === 'framebuffer' ||
                    reglType === 'framebufferCube') {
-          check(value.color.length > 0,
-            'missing color attachment for framebuffer sent to uniform "' + name + '"')
+          check.command(value.color.length > 0,
+            'missing color attachment for framebuffer sent to uniform "' + name + '"', env.commandStr)
           result = createStaticDecl(function (env) {
             return env.link(value.color[0])
           })
         } else {
-          check.raise('invalid data for uniform "' + name + '"')
+          check.commandRaise('invalid data for uniform "' + name + '"', env.commandStr)
         }
       } else if (isArrayLike(value)) {
         result = createStaticDecl(function (env) {
@@ -1662,13 +1605,13 @@ module.exports = function reglCore (
               check.command(
                 typeof value[i] === 'number' ||
                 typeof value[i] === 'boolean',
-                'invalid uniform ' + name)
+                'invalid uniform ' + name, env.commandStr)
               return value[i]
             }), ']')
           return ITEM
         })
       } else {
-        check.commandRaise('invalid or missing data for uniform "' + name + '"')
+        check.commandRaise('invalid or missing data for uniform "' + name + '"', env.commandStr)
       }
       result.value = value
       UNIFORMS[name] = result
@@ -1684,7 +1627,7 @@ module.exports = function reglCore (
     return UNIFORMS
   }
 
-  function parseAttributes (attributes) {
+  function parseAttributes (attributes, env) {
     var staticAttributes = attributes.static
     var dynamicAttributes = attributes.dynamic
 
@@ -1708,7 +1651,7 @@ module.exports = function reglCore (
           record.type = buffer.dtype
         } else {
           check.command(typeof value === 'object' && value,
-            'invalid data for attribute ' + attribute)
+            'invalid data for attribute ' + attribute, env.commandStr)
           if (value.constant) {
             var constant = value.constant
             record.buffer = 'null'
@@ -1720,7 +1663,7 @@ module.exports = function reglCore (
                 isArrayLike(constant) &&
                 constant.length > 0 &&
                 constant.length <= 4,
-                'invalid constant for attribute ' + attribute)
+                'invalid constant for attribute ' + attribute, env.commandStr)
               CUTE_COMPONENTS.forEach(function (c, i) {
                 if (i < constant.length) {
                   record[c] = constant[i]
@@ -1734,19 +1677,19 @@ module.exports = function reglCore (
             } else {
               buffer = bufferState.getBuffer(value.buffer)
             }
-            check.command(!!buffer, 'missing buffer for attribute "' + attribute + '"')
+            check.command(!!buffer, 'missing buffer for attribute "' + attribute + '"', env.commandStr)
 
             var offset = value.offset | 0
             check.command(offset >= 0,
-              'invalid offset for attribute "' + attribute + '"')
+              'invalid offset for attribute "' + attribute + '"', env.commandStr)
 
             var stride = value.stride | 0
             check.command(stride >= 0 && stride < 256,
-              'invalid stride for attribute "' + attribute + '", must be integer betweeen [0, 255]')
+              'invalid stride for attribute "' + attribute + '", must be integer betweeen [0, 255]', env.commandStr)
 
             var size = value.size | 0
             check.command(!('size' in value) || (size > 0 && size <= 4),
-              'invalid size for attribute "' + attribute + '", must be 1,2,3,4')
+              'invalid size for attribute "' + attribute + '", must be 1,2,3,4', env.commandStr)
 
             var normalized = !!value.normalized
 
@@ -1754,20 +1697,20 @@ module.exports = function reglCore (
             if ('type' in value) {
               check.commandParameter(
                 value.type, glTypes,
-                'invalid type for attribute ' + attribute)
+                'invalid type for attribute ' + attribute, env.commandStr)
               type = glTypes[value.type]
             }
 
             var divisor = value.divisor | 0
             if ('divisor' in value) {
               check.command(divisor === 0 || extInstancing,
-                'cannot specify divisor for attribute "' + attribute + '", instancing not supported')
+                'cannot specify divisor for attribute "' + attribute + '", instancing not supported', env.commandStr)
               check.command(divisor >= 0,
-                'invalid divisor for attribute "' + attribute + '"')
+                'invalid divisor for attribute "' + attribute + '"', env.commandStr)
             }
 
             check.optional(function () {
-              var command = check.guessCommand()
+              var command = env.commandStr
 
               var VALID_KEYS = [
                 'buffer',
@@ -1938,13 +1881,65 @@ module.exports = function reglCore (
     return result
   }
 
-  function parseArguments (options, attributes, uniforms, context) {
-    var result = parseOptions(options)
+  function parseArguments (options, attributes, uniforms, context, env) {
+    var staticOptions = options.static
+    var dynamicOptions = options.dynamic
 
-    result.profile = parseProfile(options)
-    result.uniforms = parseUniforms(uniforms)
-    result.attributes = parseAttributes(attributes)
-    result.context = parseContext(context)
+    check.optional(function () {
+      var KEY_NAMES = [
+        S_FRAMEBUFFER,
+        S_VERT,
+        S_FRAG,
+        S_ELEMENTS,
+        S_PRIMITIVE,
+        S_OFFSET,
+        S_COUNT,
+        S_INSTANCES,
+        S_PROFILE
+      ].concat(GL_STATE_NAMES)
+
+      function checkKeys (dict) {
+        Object.keys(dict).forEach(function (key) {
+          check.command(
+            KEY_NAMES.indexOf(key) >= 0,
+            'unknown parameter "' + key + '"',
+            env.commandStr)
+        })
+      }
+
+      checkKeys(staticOptions)
+      checkKeys(dynamicOptions)
+    })
+
+    var framebuffer = parseFramebuffer(options, env)
+    var viewportAndScissor = parseViewportScissor(options, framebuffer, env)
+    var draw = parseDraw(options, env)
+    var state = parseGLState(options, env)
+    var shader = parseProgram(options, env)
+
+    function copyBox (name) {
+      var defn = viewportAndScissor[name]
+      if (defn) {
+        state[name] = defn
+      }
+    }
+    copyBox(S_VIEWPORT)
+    copyBox(propName(S_SCISSOR_BOX))
+
+    var dirty = Object.keys(state).length > 0
+
+    var result = {
+      framebuffer: framebuffer,
+      draw: draw,
+      shader: shader,
+      state: state,
+      dirty: dirty
+    }
+
+    result.profile = parseProfile(options, env)
+    result.uniforms = parseUniforms(uniforms, env)
+    result.attributes = parseAttributes(attributes, env)
+    result.context = parseContext(context, env)
     return result
   }
 
@@ -2368,7 +2363,7 @@ module.exports = function reglCore (
           var value = arg.value
           check.command(
             value !== null && typeof value !== 'undefined',
-            'missing uniform "' + name + '"')
+            'missing uniform "' + name + '"', env.commandStr)
           if (type === GL_SAMPLER_2D || type === GL_SAMPLER_CUBE) {
             check.command(
               typeof value === 'function' &&
@@ -2378,7 +2373,7 @@ module.exports = function reglCore (
               (type === GL_SAMPLER_CUBE &&
                 (value._reglType === 'textureCube' ||
                 value._reglType === 'framebufferCube'))),
-              'invalid texture for uniform ' + name)
+              'invalid texture for uniform ' + name, env.commandStr)
             var TEX_VALUE = env.link(value._texture || value.color[0]._texture)
             scope(GL, '.uniform1i(', LOCATION, ',', TEX_VALUE + '.bind());')
             scope.exit(TEX_VALUE, '.unbind();')
@@ -2388,12 +2383,12 @@ module.exports = function reglCore (
             type === GL_FLOAT_MAT4) {
             check.optional(function () {
               check.command(isArrayLike(value),
-                'invalid matrix for uniform ' + name)
+                'invalid matrix for uniform ' + name, env.commandStr)
               check.command(
                 (type === GL_FLOAT_MAT2 && value.length === 4) ||
                 (type === GL_FLOAT_MAT3 && value.length === 9) ||
                 (type === GL_FLOAT_MAT4 && value.length === 16),
-                'invalid length for matrix uniform ' + name)
+                'invalid length for matrix uniform ' + name, env.commandStr)
             })
             var MAT_VALUE = env.global.def('new Float32Array([' +
               Array.prototype.slice.call(value) + '])')
@@ -2409,69 +2404,69 @@ module.exports = function reglCore (
           } else {
             switch (type) {
               case GL_FLOAT:
-                check.commandType(value, 'number', 'uniform ' + name)
+                check.commandType(value, 'number', 'uniform ' + name, env.commandStr)
                 infix = '1f'
                 break
               case GL_FLOAT_VEC2:
                 check.command(
                   isArrayLike(value) && value.length === 2,
-                  'uniform ' + name)
+                  'uniform ' + name, env.commandStr)
                 infix = '2f'
                 break
               case GL_FLOAT_VEC3:
                 check.command(
                   isArrayLike(value) && value.length === 3,
-                  'uniform ' + name)
+                  'uniform ' + name, env.commandStr)
                 infix = '3f'
                 break
               case GL_FLOAT_VEC4:
                 check.command(
                   isArrayLike(value) && value.length === 4,
-                  'uniform ' + name)
+                  'uniform ' + name, env.commandStr)
                 infix = '4f'
                 break
               case GL_BOOL:
-                check.commandType(value, 'boolean', 'uniform ' + name)
+                check.commandType(value, 'boolean', 'uniform ' + name, env.commandStr)
                 infix = '1i'
                 break
               case GL_INT:
-                check.commandType(value, 'number', 'uniform ' + name)
+                check.commandType(value, 'number', 'uniform ' + name, env.commandStr)
                 infix = '1i'
                 break
               case GL_BOOL_VEC2:
                 check.command(
                   isArrayLike(value) && value.length === 2,
-                  'uniform ' + name)
+                  'uniform ' + name, env.commandStr)
                 infix = '2i'
                 break
               case GL_INT_VEC2:
                 check.command(
                   isArrayLike(value) && value.length === 2,
-                  'uniform ' + name)
+                  'uniform ' + name, env.commandStr)
                 infix = '2i'
                 break
               case GL_BOOL_VEC3:
                 check.command(
                   isArrayLike(value) && value.length === 3,
-                  'uniform ' + name)
+                  'uniform ' + name, env.commandStr)
                 infix = '3i'
                 break
               case GL_INT_VEC3:
                 check.command(
                   isArrayLike(value) && value.length === 3,
-                  'uniform ' + name)
+                  'uniform ' + name, env.commandStr)
                 infix = '3i'
                 break
               case GL_BOOL_VEC4:
                 check.command(
                   isArrayLike(value) && value.length === 4,
-                  'uniform ' + name)
+                  'uniform ' + name, env.commandStr)
                 infix = '4i'
                 break
               case GL_INT_VEC4:
                 check.command(
                   isArrayLike(value) && value.length === 4,
-                  'uniform ' + name)
+                  'uniform ' + name, env.commandStr)
                 infix = '4i'
                 break
             }
@@ -2518,7 +2513,7 @@ module.exports = function reglCore (
         function checkVector (n, type) {
           check(
             shared.isArrayLike + '(' + VALUE + ')&&' + VALUE + '.length===' + n,
-            'invalid vector, should have length ' + n)
+            'invalid vector, should have length ' + n, env.commandStr)
         }
 
         function checkTexture (target) {
@@ -2526,7 +2521,7 @@ module.exports = function reglCore (
             'typeof ' + VALUE + '==="function"&&' +
             VALUE + '._reglType==="texture' +
             (target === GL_TEXTURE_2D ? '2d' : 'Cube') + '"',
-            'invalid texture type')
+            'invalid texture type', env.commandStr)
         }
 
         switch (type) {
@@ -3253,7 +3248,7 @@ module.exports = function reglCore (
       splatObject(env, options, name)
     })
 
-    var args = parseArguments(options, attributes, uniforms, context)
+    var args = parseArguments(options, attributes, uniforms, context, env)
 
     emitDrawProc(env, args)
     emitScopeProc(env, args)

